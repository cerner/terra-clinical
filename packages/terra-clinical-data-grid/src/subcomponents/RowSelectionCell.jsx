import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';

import ThemeContext from 'terra-theme-context';
import Cell from './Cell';

import styles from './RowSelectionCell.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * String identifier of the section in which the Cell will be rendered.
   */
  sectionId: PropTypes.string.isRequired,
  /**
   * String identifer of the row in which the Cell will be rendered.
   */
  rowId: PropTypes.string.isRequired,
  /**
   * String identifier of the column in which the RowSelectionCell will be rendered.
   */
  columnId: PropTypes.string.isRequired,
  /**
   * String-formatted width that the RowSelectionCell should be rendered as. Any valid css width value is supported (i.e. 200px, 3rem).
   */
  width: PropTypes.string.isRequired,
  /**
   * Boolean indicating whether or not the RowSelectionCell is selectable.
   */
  isSelectable: PropTypes.bool,
  /**
   * Boolean indicating whether the RowSelectionCell is actively selected.
   */
  isSelected: PropTypes.bool,
  /**
   * Function that will be called when the RowSelectionCell is selected. The `isSelectable` prop must be true for
   * this function to be called.
   */
  onSelect: PropTypes.func,
  /**
   * Function that will be called upon the mouse entering the selectable region of the RowSelectionCell. The `isSelectable` prop must be true for this function to be called.
   * Parameters: `onHoverStart(event)`
   */
  onHoverStart: PropTypes.func,
  /**
   * Function that will be called upon the mouse leaving the selectable region of the RowSelectionCell. The `isSelectable` prop must be true for this function to be called.
   * Parameters: `onHoverEnd(event)`
   */
  onHoverEnd: PropTypes.func,
  /**
   * Label for RowSelectionCell contents.
   */
  ariaLabel: PropTypes.string,
  /**
   * Function that will be called with a ref to the RowSelectionCell's selectable element. Parameters: `selectableRefCallback(selectableRef)`
   */
  selectableRefCallback: PropTypes.func,
};

class RowSelectionCell extends React.Component {
  constructor(props) {
    super(props);

    this.handleCellSelect = this.handleCellSelect.bind(this);
  }

  handleCellSelect(sectionId, rowId) {
    const { onSelect } = this.props;

    if (onSelect) {
      onSelect(sectionId, rowId);
    }
  }

  render() {
    const {
      sectionId,
      rowId,
      columnId,
      isSelectable,
      isSelected,
      width,
      selectableRefCallback,
      onHoverStart,
      onHoverEnd,
      ariaLabel,
    } = this.props;
    const theme = this.context;

    return (
      <Cell
        className={cx({ selectable: isSelectable, selected: isSelected }, theme.className)}
        sectionId={sectionId}
        rowId={rowId}
        columnId={columnId}
        width={width}
        isSelectable={isSelectable}
        onSelect={this.handleCellSelect}
        selectableRefCallback={selectableRefCallback}
        aria-hidden={!isSelectable || undefined}
        onHoverStart={onHoverStart}
        onHoverEnd={onHoverEnd}
        ariaLabel={ariaLabel}
      >
        <div className={cx('icon-container')}>
          <div className={cx('icon')} />
        </div>
      </Cell>
    );
  }
}

RowSelectionCell.propTypes = propTypes;
RowSelectionCell.contextType = ThemeContext;

export default RowSelectionCell;
