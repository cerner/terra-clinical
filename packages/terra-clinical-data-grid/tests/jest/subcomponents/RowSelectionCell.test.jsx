import React from 'react';
import ThemeContextProvider from 'terra-theme-context/lib/ThemeContextProvider';
import RowSelectionCell from '../../../src/subcomponents/RowSelectionCell';

describe('Cell Snapshot Tests', () => {
  it('should render a Cell with minimal props', () => {
    const cell = shallow((
      <RowSelectionCell
        sectionId="section-test"
        rowId="row-test"
        columnId="column-test"
        width="123px"
      />
    ));

    expect(cell).toMatchSnapshot();
  });

  it('should render a Cell with all provided props', () => {
    const cell = shallow((
      <RowSelectionCell
        sectionId="section-test"
        rowId="row-test"
        columnId="column-test"
        width="123px"
        isSelectable
        isSelected
        onSelect={() => {}}
        onHoverStart={() => {}}
        onHoverEnd={() => {}}
        selectableRefCallback={() => {}}
        ariaLabel="test label"
      />
    ));

    expect(cell).toMatchSnapshot();
  });

  it('correctly applies the theme context className', () => {
    const cell = mount(
      <ThemeContextProvider theme={{ className: 'orion-fusion-theme' }}>
        <RowSelectionCell
          sectionId="section-test"
          rowId="row-test"
          columnId="column-test"
          width="123px"
        />
      </ThemeContextProvider>,
    );
    expect(cell).toMatchSnapshot();
  });
});
