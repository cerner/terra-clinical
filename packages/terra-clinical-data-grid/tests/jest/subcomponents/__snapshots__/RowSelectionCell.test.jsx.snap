// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Snapshot Tests correctly applies the theme context className 1`] = `
<ThemeContextProvider
  theme={
    Object {
      "className": "orion-fusion-theme",
    }
  }
>
  <RowSelectionCell
    columnId="column-test"
    rowId="row-test"
    sectionId="section-test"
    width="123px"
  >
    <Cell
      aria-hidden={true}
      className="orion-fusion-theme"
      columnId="column-test"
      onSelect={[Function]}
      rowId="row-test"
      sectionId="section-test"
      width="123px"
    >
      <div
        aria-hidden={true}
        className="container orion-fusion-theme orion-fusion-theme"
        style={
          Object {
            "width": "123px",
          }
        }
      >
        <div
          className="content"
          role="button"
        >
          <div
            className="icon-container"
          >
            <div
              className="icon"
            />
          </div>
        </div>
      </div>
    </Cell>
  </RowSelectionCell>
</ThemeContextProvider>
`;

exports[`Cell Snapshot Tests should render a Cell with all provided props 1`] = `
<Cell
  ariaLabel="test label"
  className="selectable selected"
  columnId="column-test"
  isSelectable={true}
  onHoverEnd={[Function]}
  onHoverStart={[Function]}
  onSelect={[Function]}
  rowId="row-test"
  sectionId="section-test"
  selectableRefCallback={[Function]}
  width="123px"
>
  <div
    className="icon-container"
  >
    <div
      className="icon"
    />
  </div>
</Cell>
`;

exports[`Cell Snapshot Tests should render a Cell with minimal props 1`] = `
<Cell
  aria-hidden={true}
  className=""
  columnId="column-test"
  onSelect={[Function]}
  rowId="row-test"
  sectionId="section-test"
  width="123px"
>
  <div
    className="icon-container"
  >
    <div
      className="icon"
    />
  </div>
</Cell>
`;
